README.md
<img src="https://user-images.githubusercontent.com/106826/43119494-78be9224-8ecb-11e8-9d1a-9fc6f3014b91.png" width="300" alt="API Sprout"/>

[![Go Report Card](https://goreportcard.com/badge/github.com/puppetlabs/apisprout)](https://goreportcard.com/report/github.com/puppetlabs/apisprout) [![Build Status](https://travis-ci.com/puppetlabs/apisprout.svg?branch=master)](https://travis-ci.com/puppetlabs/apisprout) [![GitHub tag (latest SemVer)](https://img.shields.io/github/tag/puppetlabs/apisprout.svg)](https://github.com/puppetlabs/apisprout/releases)

A simple, quick, cross-platform API mock server that returns examples specified in an API description document. Features include:

- OpenAPI 3.x support
  - Uses operation `examples` or generates examples from `schema`
- Load from a URL or local file (auto reload with `--watch`)
- CORS headers enabled by default
- Accept header content negotiation
  - Example: `Accept: application/*`
- Prefer header to select response to test specific cases
  - Example: `Prefer: status=409`
- Server validation (enabled with `--validate-server`)
  - Validates scheme, hostname/port, and base path
  - Supports `localhost` out of the box
  - Use the `--add-server` flag, in conjunction with `--validate-server`, to dynamically include more servers in the validation logic
- Request parameter & body validation (enabled with `--validate-request`)
- Configuration via:
  - Files (`/etc/apisprout/config.json|yaml`)
  - Environment (prefixed with `SPROUT_`, e.g. `SPROUT_VALIDATE_SERVER`)
  - Commandline flags

Usage is simple:

```sh
# Load from a local file
apisprout my-api.yaml

# Validate server name and use base path
apisprout --validate-server my-api.yaml

# Dynamically Include a new server / path in the validation
apisprout --add-server http://localhost:8080/mock --validate-server my-api.yaml

# Load from a URL
apisprout https://raw.githubusercontent.com/OAI/OpenAPI-Specification/master/examples/v3.0/api-with-examples.yaml
```

## Docker Image

A hosted [API Sprout Docker image](https://console.cloud.google.com/gcr/images/nebula-contrib/GLOBAL/apisprout) is provided that makes it easy to deploy mock servers or run locally. For example:

```sh
docker pull gcr.io/nebula-contrib/apisprout
docker run -p 8000:8000 gcr.io/nebula-contrib/apisprout http://example.com/my-api.yaml
```

Configuration can be passed via environment variables, e.g. setting `SPROUT_VALIDATE_REQUEST=1`, or by passing commandline flags. It is also possible to use a local API description file via [Docker Volumes](https://docs.docker.com/storage/volumes/):

```
# Remember to put the full path to local archive YAML in -v
docker run -p 8000:8000 -v $FULLPATH/localfile.yaml:/api.yaml gcr.io/nebula-contrib/apisprout /api.yaml
```

## Installation

Download the appropriate binary from the [releases](https://github.com/puppetlabs/apisprout/releases) page.

Alternatively, you can use `go get`:

```sh
go get github.com/puppetlabs/apisprout
```

## Extra Features

### Remote Reload

If your API spec is loaded from a remote URL, you can live-reload it by hitting the `/__reload` endpoint.

### Health Check

A simple endpoint which returns status code `200` is available at `/__health`. This endpoint successfully returns `200` even if `--validate-server` is turned on, and the endpoint is being accessed from a non-validated host.

## Contributing

Contributions are very welcome. Please open a tracking issue or pull request and we can work to get things merged in.

## Release Process

We use [semantic-release](https://github.com/semantic-release/semantic-release) with the [ESLint convention](https://github.com/conventional-changelog/conventional-changelog/tree/master/packages/conventional-changelog-eslint) to automatically semantically version this package. A push to master creates the appropriate release.

## License

Copyright &copy; 2018-2019 Daniel G. Taylor

http://dgt.mit-license.org/


CHANGELOG.md
# [1.7.0](https://github.com/puppetlabs/apisprout/compare/v1.6.1...v1.7.0) (2019-10-29)


### Update

* Integrate upstream changes ([38a1e2ff3b84fcae46eb32562b506a0932cd0283](https://github.com/puppetlabs/apisprout/commit/38a1e2ff3b84fcae46eb32562b506a0932cd0283))

## [1.6.1](https://github.com/puppetlabs/apisprout/compare/v1.6.0...v1.6.1) (2019-08-21)


### Fix

* Correctly process sibling references to the same schema ([2487a057d41e4c9eb065f7dacb85b9f6afb57c52](https://github.com/puppetlabs/apisprout/commit/2487a057d41e4c9eb065f7dacb85b9f6afb57c52))

# [1.6.0](https://github.com/puppetlabs/apisprout/compare/v1.5.3...v1.6.0) (2019-08-21)


### Update

* Allow statuses sent with Prefer to select the default response ([7fbe771bcbb16e96e3245d3c71a9cc21bbc4c42b](https://github.com/puppetlabs/apisprout/commit/7fbe771bcbb16e96e3245d3c71a9cc21bbc4c42b))

## [1.5.3](https://github.com/puppetlabs/apisprout/compare/v1.5.2...v1.5.3) (2019-08-21)


### Fix

* Handle recursive schemas correctly ([25dcf4aff50c60017d5f11f5a38fcd77f096eb9c](https://github.com/puppetlabs/apisprout/commit/25dcf4aff50c60017d5f11f5a38fcd77f096eb9c))

## [1.5.2](https://github.com/puppetlabs/apisprout/compare/v1.5.1...v1.5.2) (2019-08-19)


### Fix

* Correct parsing of application/json, application/yaml, and text/yaml media types ([e80268d7311214901f1964d63e54891381a2344e](https://github.com/puppetlabs/apisprout/commit/e80268d7311214901f1964d63e54891381a2344e))

## [1.5.1](https://github.com/puppetlabs/apisprout/compare/v1.5.0...v1.5.1) (2019-08-18)


### Fix

* Use the correct version number in releases ([873e042b6c3d455613dbfe21eee32e7094048b83](https://github.com/puppetlabs/apisprout/commit/873e042b6c3d455613dbfe21eee32e7094048b83))

# [1.5.0](https://github.com/puppetlabs/apisprout/compare/v1.4.0...v1.5.0) (2019-08-18)


### New

* Add support for allOf, anyOf, and oneOf ([4a95528cadbd174f1b4d32dc28d4ce98ef738d1e](https://github.com/puppetlabs/apisprout/commit/4a95528cadbd174f1b4d32dc28d4ce98ef738d1e))

# [1.4.0](https://github.com/puppetlabs/apisprout/compare/v1.3.0...v1.4.0) (2019-08-18)


### Build

* Automatically release ([5ba016737ed1e3cd3feae19540e2f396b9b63533](https://github.com/puppetlabs/apisprout/commit/5ba016737ed1e3cd3feae19540e2f396b9b63533))

### Update

* Match all relevant JSON and YAML media types ([e046dd2a99faffd5b3d365459c980393913f610b](https://github.com/puppetlabs/apisprout/commit/e046dd2a99faffd5b3d365459c980393913f610b))

# [1.3.0] - 2019-03-18
- Add `--add-server` to add a custom server when using `--validate-server`.
  This allows quickly adding a custom domain or base path that will properly
  validate.
- Add `--header` (short `-H`) option to specify a custom header when fetching
  the API document. This allows you to pass custom auth info.
- Add `readOnly` and `writeOnly` support to the example generator.
- Revamped support for `--validate-server` (short `-s`)
  - Requires the use of server base path(s) on the client.
  - Localhost is now always allowed on all known base paths.
  - Support for proxy headers (e.g. `X-Forwarded-Host`).
- Better support for resolving relative path references.
- Be more resilient to parser panics when using `--watch`
- Update Docker build to use Go 1.12 and Go modules.
- Enhance example-from-schema generation code. Support enums, string formats,
  array and object examples, min/max and min items.

# [1.2.0] - 2019-02-27
- Add support for reloading OpenAPI URLs via `/__reload` on the server.
- Support external references in OpenAPI loader.
- Update dependencies, simplify file loading.
- Support jsonapi.org content type (`application/vnd.api+json`).
- Switch from `dep` to Go modules.

# [1.1.1] - 2019-01-30
- Fix `OPTIONS` request to also include CORS headers.

# [1.1.0] - 2019-01-29
- Added the `--watch` (short `-w`) parameter to reload whenever the input file
  changes. This currently only works when using files on disk.
- Update Docker build to use Go 1.11.
- Generate examples from schema when no example is available.
- Fix path parameter validation.
- Add CORS headers. Disable with `--disable-cors`.
- Documentation updates.

# [1.0.1] - 2018-10-03
- Dependency updates, fixes string format validation bug.

# [1.0.0] - 2018-07-24
- Initial release.


apisprout_test.go
package main

import (
	"fmt"
	"net/http"
	"net/http/httptest"
	"testing"

	"github.com/getkin/kin-openapi/openapi3"
	"github.com/spf13/viper"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

var localServerTests = []struct {
	name string
	in   []string
	out  []string
}{
	{
		"No servers",
		[]string{},
		[]string{},
	},
	{
		"Same path",
		[]string{
			"https://api.example.com/v1",
			"https://beta.api.example.com/v1",
		},
		[]string{
			"https://api.example.com/v1",
			"https://beta.api.example.com/v1",
			"http://localhost:8000/v1",
		},
	},
	{
		"Includes localhost already",
		[]string{
			"https://api.example.com/v1",
			"http://localhost:8000/v1",
		},
		[]string{
			"https://api.example.com/v1",
			"http://localhost:8000/v1",
		},
	},
	{
		"Invalid URL",
		[]string{
			"http://192.168.0.%31/",
		},
		[]string{},
	},
}

func TestAddLocalServers(t *testing.T) {
	viper.SetDefault("port", 8000)
	for _, tt := range localServerTests {
		t.Run(tt.name, func(t *testing.T) {
			servers := make([]*openapi3.Server, len(tt.in))
			for i, u := range tt.in {
				servers[i] = &openapi3.Server{
					URL: u,
				}
			}

			s := &openapi3.Swagger{
				Servers: servers,
			}

			err := addLocalServers(s)
			if len(tt.in) > 0 && len(tt.out) == 0 {
				assert.Error(t, err)
				return
			}
			assert.NoError(t, err)

			results := make([]string, 0, len(tt.out))
			for _, server := range s.Servers {
				results = append(results, server.URL)
			}

			assert.Equal(t, tt.out, results)
		})
	}
}

func TestMediaTypes(t *testing.T) {
	const schema = `{
		"paths": {
			"/test": {
				"get": {
					"summary": "Test",
					"responses": {
						"200": {
							"content": {
								"%s": {
									"schema": {
										type": "boolean",
										"example": true
									}
								}
							}
						}
					}
				}
			}
		}
	}`

	tests := []struct {
		MediaType  string
		StatusCode int
	}{
		{
			MediaType:  "application/json",
			StatusCode: http.StatusOK,
		},
		{
			MediaType:  "application/vnd.test-api+json",
			StatusCode: http.StatusOK,
		},
		{
			MediaType:  "application/yaml",
			StatusCode: http.StatusOK,
		},
		{
			MediaType:  "application/x-yaml",
			StatusCode: http.StatusOK,
		},
		{
			MediaType:  "application/vnd.test-api+yaml",
			StatusCode: http.StatusOK,
		},
		{
			MediaType:  "text/yaml",
			StatusCode: http.StatusOK,
		},
		{
			MediaType:  "text/x-yaml",
			StatusCode: http.StatusOK,
		},
		{
			MediaType:  "text/vnd.test-api+yaml",
			StatusCode: http.StatusOK,
		},
		{
			MediaType:  "text/vnd.test-api+xml",
			StatusCode: http.StatusInternalServerError,
		},
		{
			MediaType:  "application/json-with-extensions",
			StatusCode: http.StatusInternalServerError,
		},
	}
	for _, test := range tests {
		t.Run(test.MediaType, func(t *testing.T) {
			_, router, err := load("file:///swagger.json", []byte(fmt.Sprintf(schema, test.MediaType)))
			require.NoError(t, err)
			require.NotNil(t, router)

			rr := NewRefreshableRouter()
			rr.Set(router)

			req, err := http.NewRequest("GET", "/test", nil)
			require.NoError(t, err)

			resp := httptest.NewRecorder()
			handler(rr).ServeHTTP(resp, req)

			assert.Equal(t, test.StatusCode, resp.Code)
		})
	}
}


apisprout.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"log"
	"math/rand"
	"mime"
	"net/http"
	"net/url"
	"os"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"
	"time"

	"github.com/fsnotify/fsnotify"
	"github.com/getkin/kin-openapi/openapi3"
	"github.com/getkin/kin-openapi/openapi3filter"
	"github.com/gobwas/glob"
	"github.com/pkg/errors"
	"github.com/spf13/cobra"
	"github.com/spf13/pflag"
	"github.com/spf13/viper"
	yaml "gopkg.in/yaml.v2"
)

// GitSummary is filled in by `govvv` for version info.
var GitSummary string

// Version is filled in by `govvv` for version info.
var Version string

var (
	// ErrNoExample is sent when no example was found for an operation.
	ErrNoExample = errors.New("No example found")

	// ErrRecursive is when a schema is impossible to represent because it infinitely recurses.
	ErrRecursive = errors.New("Recursive schema")

	// ErrCannotMarshal is set when an example cannot be marshalled.
	ErrCannotMarshal = errors.New("Cannot marshal example")

	// ErrMissingAuth is set when no authorization header or key is present but
	// one is required by the API description.
	ErrMissingAuth = errors.New("Missing auth")
)

var (
	marshalJSONMatcher = regexp.MustCompile(`^application/(vnd\..+\+)?json$`)
	marshalYAMLMatcher = regexp.MustCompile(`^(application|text)/(x-|vnd\..+\+)?yaml$`)
)

type RefreshableRouter struct {
	router *openapi3filter.Router
}

func (rr *RefreshableRouter) Set(router *openapi3filter.Router) {
	rr.router = router
}

func (rr *RefreshableRouter) Get() *openapi3filter.Router {
	return rr.router
}

func NewRefreshableRouter() *RefreshableRouter {
	return &RefreshableRouter{}
}

// ContentNegotiator is used to match a media type during content negotiation
// of HTTP requests.
type ContentNegotiator struct {
	globs []glob.Glob
}

// NewContentNegotiator creates a new negotiator from an HTTP Accept header.
func NewContentNegotiator(accept string) *ContentNegotiator {
	// The HTTP Accept header is parsed and converted to simple globs, which
	// can be used to match an incoming mimetype. Example:
	// Accept: text/html, text/*;q=0.9, */*;q=0.8
	// Will be turned into the following globs:
	// - text/html
	// - text/*
	// - */*
	globs := make([]glob.Glob, 0)
	for _, mt := range strings.Split(accept, ",") {
		parsed, _, _ := mime.ParseMediaType(mt)
		globs = append(globs, glob.MustCompile(parsed))
	}

	return &ContentNegotiator{
		globs: globs,
	}
}

// Match returns true if the given mediatype string matches any of the allowed
// types in the accept header.
func (cn *ContentNegotiator) Match(mediatype string) bool {
	for _, glob := range cn.globs {
		if glob.Match(mediatype) {
			return true
		}
	}

	return false
}

func main() {
	rand.Seed(time.Now().UnixNano())

	// Load configuration from file(s) if provided.
	viper.SetConfigName("config")
	viper.AddConfigPath("/etc/apisprout/")
	viper.AddConfigPath("$HOME/.apisprout/")
	viper.ReadInConfig()

	// Load configuration from the environment if provided. Flags below get
	// transformed automatically, e.g. `foo-bar` -> `SPROUT_FOO_BAR`.
	viper.SetEnvPrefix("SPROUT")
	viper.SetEnvKeyReplacer(strings.NewReplacer("-", "_"))
	viper.AutomaticEnv()

	version := Version
	if version == "" {
		version = GitSummary
	}

	// Build the root command. This is the application's entry point.
	cmd := filepath.Base(os.Args[0])
	root := &cobra.Command{
		Use:     fmt.Sprintf("%s [flags] FILE", cmd),
		Version: version,
		Args:    cobra.MinimumNArgs(1),
		Run:     server,
		Example: fmt.Sprintf("  # Basic usage\n  %s openapi.yaml\n\n  # Validate server name and use base path\n  %s --validate-server openapi.yaml\n\n  # Fetch API via HTTP with custom auth header\n  %s -H 'Authorization: abc123' http://example.com/openapi.yaml", cmd, cmd, cmd),
	}

	// Set up global options.
	flags := root.PersistentFlags()

	addParameter(flags, "port", "p", 8000, "HTTP port")
	addParameter(flags, "validate-server", "s", false, "Check scheme/hostname/basepath against configured servers")
	addParameter(flags, "validate-request", "", false, "Check request data structure")
	addParameter(flags, "watch", "w", false, "Reload when input file changes")
	addParameter(flags, "disable-cors", "", false, "Disable CORS headers")
	addParameter(flags, "header", "H", "", "Add a custom header when fetching API")
	addParameter(flags, "add-server", "", "", "Add a new valid server URL, use with --validate-server")
	addParameter(flags, "https", "", false, "Use HTTPS instead of HTTP")
	addParameter(flags, "public-key", "", "", "Public key for HTTPS, use with --https")
	addParameter(flags, "private-key", "", "", "Private key for HTTPS, use with --https")

	// Run the app!
	root.Execute()
}

// addParameter adds a new global parameter with a default value that can be
// configured using configuration files, the environment, or commandline flags.
func addParameter(flags *pflag.FlagSet, name, short string, def interface{}, desc string) {
	viper.SetDefault(name, def)
	switch v := def.(type) {
	case bool:
		flags.BoolP(name, short, v, desc)
	case int:
		flags.IntP(name, short, v, desc)
	case string:
		flags.StringP(name, short, v, desc)
	}
	viper.BindPFlag(name, flags.Lookup(name))
}

// getTypedExample will return an example from a given media type, if such an
// example exists. If multiple examples are given, then one is selected at
// random unless an "example" item exists in the Prefer header
func getTypedExample(mt *openapi3.MediaType, prefer map[string]string) (interface{}, error) {
	if mt.Example != nil {
		return mt.Example, nil
	}

	if len(mt.Examples) > 0 {
		// If preferred example requested and it it exists, return it
		preferredExample := ""
		if mapContainsKey(prefer, "example") {
			preferredExample = prefer["example"]
			if _, ok := mt.Examples[preferredExample]; ok {
				return mt.Examples[preferredExample].Value.Value, nil
			}
		}

		// Choose a random example to return.
		keys := make([]string, 0, len(mt.Examples))
		for k := range mt.Examples {
			keys = append(keys, k)
		}

		if len(keys) > 0 {
			selected := keys[rand.Intn(len(keys))]
			return mt.Examples[selected].Value.Value, nil
		}
	}

	if mt.Schema != nil {
		return OpenAPIExample(ModeResponse, mt.Schema.Value)
	}
	// TODO: generate data from JSON schema, if no examples available?

	return nil, ErrNoExample
}

// getExample tries to return an example for a given operation.
// Using the Prefer http header, the consumer can specify the type of response they want.
func getExample(negotiator *ContentNegotiator, prefer map[string]string, op *openapi3.Operation) (int, string, map[string]*openapi3.HeaderRef, interface{}, error) {
	var responses []string
	var blankHeaders = make(map[string]*openapi3.HeaderRef)

	if !mapContainsKey(prefer, "status") {
		// First, make a list of responses ordered by successful (200-299 status code)
		// before other types.
		success := make([]string, 0)
		other := make([]string, 0)
		for s := range op.Responses {
			if status, err := strconv.Atoi(s); err == nil && status >= 200 && status < 300 {
				success = append(success, s)
				continue
			}
			other = append(other, s)
		}
		responses = append(success, other...)
	} else if op.Responses[prefer["status"]] != nil {
		responses = []string{prefer["status"]}
	} else if op.Responses["default"] != nil {
		responses = []string{"default"}
	} else {
		return 0, "", blankHeaders, nil, ErrNoExample
	}

	// Now try to find the first example we can and return it!
	for _, s := range responses {
		response := op.Responses[s]
		status, err := strconv.Atoi(s)
		if err != nil {
			// If we are using the default with prefer, we can use its status
			// code:
			status, err = strconv.Atoi(prefer["status"])
		}
		if err != nil {
			// Otherwise, treat default and other named statuses as 200.
			status = http.StatusOK
		}

		if response.Value.Content == nil {
			// This is a valid response but has no body defined.
			return status, "", blankHeaders, "", nil
		}

		for mt, content := range response.Value.Content {
			if negotiator != nil && !negotiator.Match(mt) {
				// This is not what the client asked for.
				continue
			}

			example, err := getTypedExample(content, prefer)
			if err == nil {
				return status, mt, response.Value.Headers, example, nil
			}

			fmt.Printf("Error getting example: %v\n", err)
		}
	}

	return 0, "", blankHeaders, nil, ErrNoExample
}

// addLocalServers will ensure that requests to localhost are always allowed
// even if not specified in the OpenAPI document.
func addLocalServers(swagger *openapi3.Swagger) error {
	seen := make(map[string]bool)
	for _, s := range swagger.Servers {
		seen[s.URL] = true
	}

	lservers := make([]*openapi3.Server, 0, len(swagger.Servers))
	for _, s := range swagger.Servers {
		u, err := url.Parse(s.URL)
		if err != nil {
			return err
		}

		if u.Hostname() != "localhost" {
			u.Scheme = "http"
			u.Host = fmt.Sprintf("localhost:%d", viper.GetInt("port"))

			ls := &openapi3.Server{
				URL:         u.String(),
				Description: s.Description,
				Variables:   s.Variables,
			}

			if !seen[ls.URL] {
				lservers = append(lservers, ls)
				seen[ls.URL] = true
			}
		}
	}

	if len(lservers) != 0 {
		swagger.Servers = append(swagger.Servers, lservers...)
	}

	return nil
}

// Load the OpenAPI document and create the router.
func load(uri string, data []byte) (swagger *openapi3.Swagger, router *openapi3filter.Router, err error) {
	defer func() {
		if r := recover(); r != nil {
			swagger = nil
			router = nil
			if e, ok := r.(error); ok {
				err = errors.Wrap(e, "Caught panic while trying to load")
			} else {
				err = fmt.Errorf("Caught panic while trying to load")
			}
		}
	}()

	loader := openapi3.NewSwaggerLoader()
	loader.IsExternalRefsAllowed = true

	var u *url.URL
	u, err = url.Parse(uri)
	if err != nil {
		return
	}

	swagger, err = loader.LoadSwaggerFromDataWithPath(data, u)
	if err != nil {
		return
	}

	if !viper.GetBool("validate-server") {
		// Clear the server list so no validation happens. Note: this has a side
		// effect of no longer parsing any server-declared parameters.
		swagger.Servers = make([]*openapi3.Server, 0)
	} else {
		// Special-case localhost to always be allowed for local testing.
		if err = addLocalServers(swagger); err != nil {
			return
		}

		if cs := viper.GetString("add-server"); cs != "" {
			swagger.Servers = append(swagger.Servers, &openapi3.Server{
				URL:         cs,
				Description: "Custom server from command line param",
				Variables:   make(map[string]*openapi3.ServerVariable),
			})
		}
	}

	// Create a new router using the OpenAPI document's declared paths.
	router = openapi3filter.NewRouter().WithSwagger(swagger)

	return
}

// parsePreferHeader takes the value of a prefer header and splits it out into key value pairs
//
// HTTP Prefer header specification examples:
// - Prefer: status=200; example="something"
// - Prefer: example=something;status=200;
// - Prefer: example="somet,;hing";status=200;
//
// As part of the Prefer specification, it is completely valid to specify
// multiple Prefer headers in a single request, however we won't be
// supporting that for the moment and only the first Prefer header
// will be used.
func parsePreferHeader(value string) map[string]string {
	prefer := map[string]string{}
	if value != "" {
		// In the event that something is quoted, we want to pull those items out of the string
		// and save them for later, so they don't conflict with other splitting logic.

		quotedRegex := regexp.MustCompile(`"[^"]*"`)
		splitRegex := regexp.MustCompile(`(,|;| )`)
		wilcardRegex := regexp.MustCompile(`%%([0-9]+)%%`)

		quotedStrings := quotedRegex.FindAllString(value, -1)
		if len(quotedStrings) > 0 {
			// replace each quoted string with a placehoder
			for idx, quotedString := range quotedStrings {
				value = strings.Replace(value, quotedString, fmt.Sprintf("%%%%%v%%%%", idx), 1)
			}
		}

		pairs := splitRegex.Split(value, -1)
		for _, pair := range pairs {
			pair = strings.TrimSpace(pair)
			if pair != "" {
				// Put any wildcards back
				wildcardStrings := wilcardRegex.FindAllStringSubmatch(pair, -1)
				for _, wildcard := range wildcardStrings {
					quotedIdx, _ := strconv.Atoi(wildcard[1])
					pair = strings.Replace(pair, wildcard[0], quotedStrings[quotedIdx], 1)
				}

				// Determine the key and valid for this argument
				if strings.Contains(pair, "=") {
					eqIdx := strings.Index(pair, "=")
					prefer[pair[:eqIdx]] = strings.Trim(pair[eqIdx+1:], `"`)
				} else {
					prefer[pair] = ""
				}
			}
		}
	}
	return prefer
}

func mapContainsKey(dict map[string]string, key string) bool {
	if _, ok := dict[key]; ok {
		return true
	}
	return false
}

var handler = func(rr *RefreshableRouter) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
		if !viper.GetBool("disable-cors") {
			corsOrigin := req.Header.Get("Origin")
			if corsOrigin == "" {
				corsOrigin = "*"
			}
			w.Header().Set("Access-Control-Allow-Origin", corsOrigin)

			if corsOrigin != "*" {
				// Allow credentials to be sent if an origin has  been specified.
				// This is done *outside* of an OPTIONS request since it might be
				// required for a non-preflighted GET/POST request.
				w.Header().Set("Access-Control-Allow-Credentials", "true")
			}

			// Handle pre-flight OPTIONS request
			if (*req).Method == "OPTIONS" {
				corsMethod := req.Header.Get("Access-Control-Request-Method")
				if corsMethod == "" {
					corsMethod = "POST, GET, OPTIONS, PUT, DELETE"
				}

				corsHeaders := req.Header.Get("Access-Control-Request-Headers")
				if corsHeaders == "" {
					corsHeaders = "Accept, Content-Type, Content-Length, Accept-Encoding, X-CSRF-Token, Authorization"
				}

				w.Header().Set("Access-Control-Allow-Methods", corsMethod)
				w.Header().Set("Access-Control-Allow-Headers", corsHeaders)
				return
			}
		}

		info := fmt.Sprintf("%s %v", req.Method, req.URL)

		// Set up the request, handling potential proxy headers
		req.URL.Host = req.Host
		fHost := req.Header.Get("X-Forwarded-Host")
		if fHost != "" {
			req.URL.Host = fHost
		}

		req.URL.Scheme = "http"
		if req.Header.Get("X-Forwarded-Proto") == "https" ||
			req.Header.Get("X-Forwarded-Scheme") == "https" ||
			strings.Contains(req.Header.Get("Forwarded"), "proto=https") {
			req.URL.Scheme = "https"
		}

		if viper.GetBool("validate-server") {
			// Use the scheme/host in the log message since we are validating it.
			info = fmt.Sprintf("%s %v", req.Method, req.URL)
		}

		route, pathParams, err := rr.Get().FindRoute(req.Method, req.URL)
		if err != nil {
			log.Printf("ERROR: %s => %v", info, err)
			w.WriteHeader(http.StatusNotFound)
			return
		}

		if viper.GetBool("validate-request") {
			err = openapi3filter.ValidateRequest(nil, &openapi3filter.RequestValidationInput{
				Request:    req,
				Route:      route,
				PathParams: pathParams,
				Options: &openapi3filter.Options{
					AuthenticationFunc: func(c context.Context, input *openapi3filter.AuthenticationInput) error {
						// TODO: support more schemes
						sec := input.SecurityScheme
						if sec.Type == "http" && sec.Scheme == "bearer" {
							if req.Header.Get("Authorization") == "" {
								return ErrMissingAuth
							}
						}
						return nil
					},
				},
			})
			if err != nil {
				log.Printf("ERROR: %s => %v", info, err)
				w.WriteHeader(http.StatusBadRequest)
				w.Write([]byte(fmt.Sprintf("%v", err)))
				return
			}
		}

		var negotiator *ContentNegotiator
		if accept := req.Header.Get("Accept"); accept != "" {
			negotiator = NewContentNegotiator(accept)
			if accept != "*/*" {
				info = fmt.Sprintf("%s (Accept %s)", info, accept)
			}
		}

		prefer := parsePreferHeader(req.Header.Get("Prefer"))

		status, mediatype, headers, example, err := getExample(negotiator, prefer, route.Operation)
		if err != nil {
			log.Printf("%s => Missing example", info)
			w.WriteHeader(http.StatusTeapot)
			w.Write([]byte("No example available."))
			return
		}

		id := route.Operation.OperationID
		if id == "" {
			id = route.Operation.Summary
		}

		log.Printf("%s (%s) => %d (%s)", info, id, status, mediatype)

		var encoded []byte

		if s, ok := example.(string); ok {
			encoded = []byte(s)
		} else if _, ok := example.([]byte); ok {
			encoded = example.([]byte)
		} else {
			if marshalJSONMatcher.MatchString(mediatype) {
				encoded, err = json.MarshalIndent(example, "", "  ")
			} else if marshalYAMLMatcher.MatchString(mediatype) {
				encoded, err = yaml.Marshal(example)
			} else {
				log.Printf("Cannot marshal as '%s'!", mediatype)
				err = ErrCannotMarshal
			}

			if err != nil {
				w.WriteHeader(http.StatusInternalServerError)
				w.Write([]byte("Unable to marshal response"))
				return
			}
		}

		for name, header := range headers {
			if header.Value != nil {
				example := name

				if header.Value.Schema != nil && header.Value.Schema.Value != nil {
					if v, err := OpenAPIExample(ModeResponse, header.Value.Schema.Value); err == nil {
						if vs, ok := v.(string); ok {
							example = vs
						} else {
							fmt.Printf("Could not convert example value '%v' to string", v)
						}
					}
				}

				w.Header().Set(name, example)
			}
		}

		if mediatype != "" {
			w.Header().Set("Content-Type", mediatype)
		}

		w.WriteHeader(status)
		w.Write(encoded)
	})
}

// server loads an OpenAPI file and runs a mock server using the paths and
// examples defined in the file.
func server(cmd *cobra.Command, args []string) {
	var swagger *openapi3.Swagger
	rr := NewRefreshableRouter()

	uri := args[0]

	var err error
	var data []byte
	dataType := strings.Trim(strings.ToLower(filepath.Ext(uri)), ".")

	// Load either from an HTTP URL or from a local file depending on the passed
	// in value.
	if strings.HasPrefix(uri, "http") {
		req, err := http.NewRequest("GET", uri, nil)
		if err != nil {
			log.Fatal(err)
		}
		if customHeader := viper.GetString("header"); customHeader != "" {
			header := strings.Split(customHeader, ":")
			if len(header) != 2 {
				log.Fatal("Header format is invalid.")
			}
			req.Header.Add(strings.TrimSpace(header[0]), strings.TrimSpace(header[1]))
		}
		client := &http.Client{}
		resp, err := client.Do(req)
		if err != nil {
			log.Fatal(err)
		}

		data, err = ioutil.ReadAll(resp.Body)
		resp.Body.Close()
		if err != nil {
			log.Fatal(err)
		}

		if viper.GetBool("watch") {
			log.Fatal("Watching a URL is not supported.")
		}
	} else {
		data, err = ioutil.ReadFile(uri)
		if err != nil {
			log.Fatal(err)
		}

		if viper.GetBool("watch") {
			// Set up a new filesystem watcher and reload the router every time
			// the file has changed on disk.
			watcher, err := fsnotify.NewWatcher()
			if err != nil {
				log.Fatal(err)
			}
			defer watcher.Close()

			go func() {
				// Since waiting for events or errors is blocking, we do this in a
				// goroutine. It loops forever here but will exit when the process
				// is finished, e.g. when you `ctrl+c` to exit.
				for {
					select {
					case event, ok := <-watcher.Events:
						if !ok {
							return
						}
						if event.Op&fsnotify.Write == fsnotify.Write {
							fmt.Printf("🌙 Reloading %s\n", uri)
							data, err = ioutil.ReadFile(uri)
							if err != nil {
								log.Fatal(err)
							}

							if s, r, err := load(uri, data); err == nil {
								swagger = s
								rr.Set(r)
							} else {
								log.Printf("ERROR: Unable to load OpenAPI document: %s", err)
							}
						}
					case err, ok := <-watcher.Errors:
						if !ok {
							return
						}
						fmt.Println("error:", err)
					}
				}
			}()

			watcher.Add(uri)
		}
	}

	swagger, router, err := load(uri, data)
	if err != nil {
		log.Fatal(err)
	}

	rr.Set(router)

	if strings.HasPrefix(uri, "http") {
		http.HandleFunc("/__reload", func(w http.ResponseWriter, r *http.Request) {
			resp, err := http.Get(uri)
			if err != nil {
				log.Printf("ERROR: %v", err)
				w.WriteHeader(http.StatusBadRequest)
				w.Write([]byte("error while reloading"))
				return
			}

			data, err = ioutil.ReadAll(resp.Body)
			resp.Body.Close()
			if err != nil {
				log.Printf("ERROR: %v", err)
				w.WriteHeader(http.StatusBadRequest)
				w.Write([]byte("error while parsing"))
				return
			}

			if s, r, err := load(uri, data); err == nil {
				swagger = s
				rr.Set(r)
			}

			w.WriteHeader(200)
			w.Write([]byte("reloaded"))
			log.Printf("Reloaded from %s", uri)
		})
	}

	// Add a health check route which returns 200
	http.HandleFunc("/__health", func(w http.ResponseWriter, r *http.Request) {
		w.WriteHeader(200)
		log.Printf("Health check")
	})

	// Another custom handler to return the exact swagger document given to us
	http.HandleFunc("/__schema", func(w http.ResponseWriter, req *http.Request) {
		if !viper.GetBool("disable-cors") {
			corsOrigin := req.Header.Get("Origin")
			if corsOrigin == "" {
				corsOrigin = "*"
			}
			w.Header().Set("Access-Control-Allow-Origin", corsOrigin)
		}

		w.Header().Set("Content-Type", fmt.Sprintf("application/%v; charset=utf-8", dataType))
		w.WriteHeader(http.StatusOK)
		fmt.Fprint(w, string(data))
	})

	// Register our custom HTTP handler that will use the router to find
	// the appropriate OpenAPI operation and try to return an example.
	http.Handle("/", handler(rr))

	format := "🌱 Sprouting %s on port %d"
	if viper.GetBool("https") {
		format = "🌱 Securely sprouting %s on port %d"
	}
	fmt.Printf(format, swagger.Info.Title, viper.GetInt("port"))

	if viper.GetBool("validate-server") && len(swagger.Servers) != 0 {
		fmt.Printf(" with valid servers:\n")
		for _, s := range swagger.Servers {
			fmt.Println("• " + s.URL)
		}
	} else {
		fmt.Printf("\n")
	}

	port := fmt.Sprintf(":%d", viper.GetInt("port"))
	if viper.GetBool("https") {
		err = http.ListenAndServeTLS(port, viper.GetString("public-key"),
			viper.GetString("private-key"), nil)
	} else {
		err = http.ListenAndServe(port, nil)
	}
	if err != nil {
		log.Fatal(err)
	}
}
